src/cdk/schematics/ng-update/devkit-file-system.ts

```diff
@@ -73,4 +62,23 @@ export class DevkitFileSystem extends FileSystem<Path> {
     const buffer = this._tree.read(filePath);
     return buffer !== null ? buffer.toString() : null;
   }
+
+  existsDirectory(dirPath: Path) {
+    // The devkit tree does not expose an API for checking whether a given
+    // directory exists. It throws a specific error though if a directory
+    // is being read as a file. We use that to check if a directory exists.
+    try {
+      this._tree.get(dirPath);
+    } catch (e) {
+      if (e instanceof PathIsDirectoryException) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  readDirectory(dirPath: Path): DirectoryEntry {
+    const {subdirs: directories, subfiles: files} = this._tree.getDir(dirPath);
```

Comment:
Not something we control, but it's a little weird for this to be called "sub files" since all files are nested in each other. Maybe we can change it if we're changing other things.

Reply:
Yeah, it's definitely a bit ambiguous. You mean that we create a proposal upstream for them to change it?